File: solana/programs/chainbills/src/state/payment.rs
use crate::state::TokenAndAmount;
use anchor_lang::prelude::*;

#[account]
pub struct Payment {
  /// The nth count of global payments at the point this payment was made.
  pub global_count: u64, // 8 bytes

  /// The nth count of payments on the calling chain at the point this payment
  /// was made.
  pub chain_count: u64, // 8 bytes

  /// The address of the Payable to which this Payment was made.
  pub payable: Pubkey, // 32 bytes

  /// The address of the User account that made this Payment.
  pub payer: Pubkey, // 32 bytes

  /// The nth count of payments that the payer has made
  /// at the point of making this payment.
  pub payer_count: u64, // 8 bytes

  /// The nth count of payments that the payable has received
  /// at the point when this payment was made.
  pub payable_count: u64, // 8 bytes

  /// When this payment was made.
  pub timestamp: u64, // 8 bytes

  /// The amount and token that the payer paid
  pub details: TokenAndAmount, // TokenAndAmount::SPACE
}

impl Payment {
  // discriminator (8) included
  pub const SPACE: usize = (6 * 8) + (2 * 32) + TokenAndAmount::SPACE;

  /// AKA `b"payment"`.
  #[constant]
  pub const SEED_PREFIX: &'static [u8] = b"payment";
}

File: server/src/schemas/payment.ts
import { Network } from '@wormhole-foundation/sdk';

import { Chain } from '../utils/chain';
import { TokenAndAmount, TokenAndAmountOffChain } from './tokens-and-amounts';

export class Payment {
  id: string;
  globalCount: number;
  chain: Chain;
  chainCount: number;
  network: Network;
  payer: string;
  payerCount: number;
  payerWallet: string;
  payable: string;
  payableCount: number;
  timestamp: number;
  details: TokenAndAmountOffChain;

  constructor(
    id: string,
    chain: Chain,
    network: Network,
    payerWallet: string,
    onChainData: any
  ) {
    this.id = id;
    this.globalCount = onChainData.globalCount.toNumber();
    this.chain = chain;
    this.chainCount = onChainData.chainCount.toNumber();
    this.network = network;
    this.payer = onChainData.payer.toBase58();
    this.payerCount = onChainData.payerCount.toNumber();
    this.payerWallet = payerWallet;
    this.payable = onChainData.payable.toBase58();
    this.payableCount = onChainData.payableCount.toNumber();
    this.details = TokenAndAmount.fromOnChain(onChainData.details, chain);
    this.timestamp = onChainData.timestamp.toNumber();
  }
}
